### PR TITLE
Web login

### PR DESCRIPTION
**Pull request by MarkSalabutin**
#### Methods
1. Request login page: 
- checks **appID** and **scopes** params
- checks if user is authenticated if so redirects to the callback URL with the IDToken in the URL
- otherwise returns the login page
2. Login with credentials: 
- checks credentials
- sets cookie that means that the user is authenticated
- redirects to the login page which will redirect the user to the callback URL
3. Renew IDToken:
- checks **appID** and **scopes** params
- checks if user is authenticated
- returns a template which will send IDToken to the app
4. Logout:
- removes cookie that means that the user is authenticated
- redirects to the login page

### CODE REVIEW

File: encryptor/encryptor.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+	"github.com/madappgang/identifo/model"
+)
+
+// EncryptitonService holds key.
+type EncryptitonService struct {
+	key []byte
+}
+
+// Encrypt encrypts plaintext with symmetric encryption method
```
Comment (esimonov) [inline encryptor/encryptor.go]: Period at the end of sentences please.

File: model/token_service.go [hunk 2]
```diff
@@ -49,7 +49,7 @@ type Token interface {
 	Payload() map[string]string
 }
 
-//Validator calidate token with external requester
+//Validator validate token with external requester
```
Comment (esimonov) [inline model/token_service.go]: Please add one space after dashes. Also, validate**s**. Also, the period at the end of a sentence is needed.

File: mongo/user_storage.go [hunk 3]
```diff
@@ -219,7 +219,7 @@ func (us *UserStorage) ImportJSON(data []byte) error {
 //data implementation
 type userData struct {
 	ID           bson.ObjectId          `bson:"_id,omitempty" json:"id,omitempty"`
-	Name         string                 `bson:"name,omitempty" json:"name,omitempty"`
+	Name         string                 `bson:"user,omitempty" json:"username,omitempty"`
```
Comment (esimonov) [inline mongo/user_storage.go]: What is the reason?

File: web/html/app_middleware.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+	"net/http"
+	"path"
+	"strings"
+
+	"github.com/madappgang/identifo/model"
+	"github.com/urfave/negroni"
+)
+
+const (
+	// FormKeyAppID form key to keep application ID.
```
Comment (esimonov) [inline web/html/app_middleware.go]: FormKeyAppID **is a** form key *that holds* application ID, I would say so.

File: web/html/login.go [hunk 5]
```diff
@@ -0,0 +1,58 @@
+package html
+
+import (
+	"encoding/json"
+	"net/http"
+	"path"
+)
+
+const usernameKey = "email"
```
Comment (esimonov) [inline web/html/login.go]: You can wrap them all in a single const statement, like this:
```go
const (
        usernameKey = "email"
        passwordKey = "password"
        scopesKey = "scopes"
)
```

File: web/html/login.go [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+		scopes, err = ar.UserStorage.RequestScopes(user.ID(), scopes)
+		if err != nil {
+			ar.Logger.Printf("Error: invalid scopes %v for userID: %v", scopes, user.ID())
+			http.Redirect(w, r, errorPath, http.StatusFound)
+			return
+		}
+
+		cipher, err := ar.Encryptor.Encrypt([]byte(user.ID()))
+
+		setCookie(w, CookieKeyUserID, string(cipher), 60*60*24*2)
```
Comment (esimonov) [inline web/html/login.go]: Probably, would be nice to declare this MaxAge as a constant.

File: mongo/user_storage.go (hunk 3 above)
Comment (MarkSalabutin) [reply mongo/user_storage.go]: According to this file: [users.json](https://github.com/MadAppGang/identifo/blob/e71ab360ec4184c729346531c9888f5f222d7c62/cmd/import/users.json#L3) user's name should be parsed from the filed **username**.

Comment (esimonov) [reply mongo/user_storage.go]: But why do you change `name` to `user`? I don't really think that `user` is an appropriate choice for the string-typed field which only holds name.

File: jwt/token_service.go [hunk 7]
```diff
@@ -355,6 +401,14 @@ func ResetTokenLifespan(lifespan int64) func(*TokenService) error {
 	}
 }
 
+// AuthTokenLifespan sets custom lifespan in seconds for the auth token
+func AuthTokenLifespan(lifespan int64) func(*TokenService) error {
```
Comment (erudenko) [inline jwt/token_service.go]: AuthToken naming

File: model/token_service.go [hunk 8]
```diff
@@ -9,6 +9,8 @@ const (
 	AccessTokenType = "access"
 	//ResetTokenType reset password token type value
 	ResetTokenType = "reset"
+	//AuthTokenType auth token type value
+	AuthTokenType = "auth"
```
Comment (erudenko) [inline model/token_service.go]: WebCookieToken

File: web/html/app_middleware.go [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+	"github.com/urfave/negroni"
+)
+
+const (
+	// FormKeyAppID is a form key to keep application ID.
+	FormKeyAppID = "appId"
+)
+
+// AppID gets app id from the request body.
+func (ar *Router) AppID() negroni.HandlerFunc {
```
Comment (erudenko) [inline web/html/app_middleware.go]: I suppose we already have auth middleware implemented before?

File: web/html/renew_idtoken.go [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+	"net/http"
+	"path"
+	"strings"
+
+	"github.com/madappgang/identifo/jwt"
+	"github.com/madappgang/identifo/model"
+)
+
+// RenewIDToken creates new id_token if user is already authenticated.
+func (ar *Router) RenewIDToken(pathComponents ...string) http.HandlerFunc {
```
Comment (erudenko) [inline web/html/renew_idtoken.go]: this function looks too complex, please extract common login to other parts

File: web/html/app_middleware.go (hunk 9 above)
Comment (MarkSalabutin) [reply web/html/app_middleware.go]: It isn't auth middleware, this middleware check's appID param and puts app in the context.

Comment (erudenko) [reply web/html/app_middleware.go]: we already have app middleware in API router. Looks like duplicating of the same logic

Comment (MarkSalabutin) [reply web/html/app_middleware.go]: Yes, you're right. I'll reuse this logic

Comment (MarkSalabutin) [reply web/html/app_middleware.go]: But it still gonna be two middlewares, because they have different sources and different ways of handling errors.

Comment (erudenko) [reply web/html/app_middleware.go]: Yes, you need to extract common logic and reuse it in both places

Comment (MarkSalabutin) [reply web/html/app_middleware.go]: Roger that!

File: web/api/appsecret.go [hunk 11]
```diff
@@ -10,6 +10,7 @@ import (
 	"net/http"
 	"strings"
 
+	"github.com/madappgang/identifo/web/shared"
```
Comment (erudenko) [inline web/api/appsecret.go]: shared is not a good naming for this kind of context
Please created 'middlewares' for this particular case

File: web/shared/parse_token.go [hunk 12]
```diff
@@ -0,0 +1,25 @@
+package shared
+
+import (
+	"github.com/madappgang/identifo/jwt"
+	"github.com/madappgang/identifo/model"
+)
+
+// ParseToken parses token and validates it.
+func ParseToken(tstr string, ts model.TokenService, tokenType string) (model.Token, error) {
```
Comment (erudenko) [inline web/shared/parse_token.go]: looks like this logic should live in jwt package

File: web/shared/app_from_context.go [hunk 13]
```diff
@@ -0,0 +1,18 @@
+package shared
+
+import (
+	"context"
+
+	"github.com/madappgang/identifo/model"
+)
+
+// AppFromContext returns app data from request conntext.
+func AppFromContext(ctx context.Context) model.AppData {
```
Comment (erudenko) [inline web/shared/app_from_context.go]: please move to 'web/middleware'  package

File: web/shared/app_by_id.go [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+	"github.com/madappgang/identifo/model"
+)
+
+var (
+	errorInactiveApp = errors.New("App is inactive")
+	errorEmptyAppID  = errors.New("Empty appID param")
+)
+
+// AppByID gets app by id and checks if it's active.
+func AppByID(as model.AppStorage, appID string) (model.AppData, error) {
```
Comment (erudenko) [inline web/shared/app_by_id.go]: looks like this logic should be placed inside model - appStorage

File: web/html/static_files.go [hunk 15]
```diff
@@ -134,3 +139,96 @@ func (ar *Router) ResetPasswordHandler(pathComponents ...string) http.HandlerFun
 		}
 	}
 }
+
+// LoginHandler serves login page or redirects to the callback_url if user is already authenticated.
+func (ar *Router) LoginHandler(pathComponents ...string) http.HandlerFunc {
```
Comment (erudenko) [inline web/html/static_files.go]: file called static_files, this handler is not looks  like static file, I believe it deserves it's own login dedicated file

File: web/html/renew_idtoken.go [hunk 16]
```diff
@@ -0,0 +1,108 @@
+package html
```
Comment (erudenko) [inline web/html/renew_idtoken.go]: please remove 'idtoken' from naming. We call it access token here. ID token is a little bit different thing. Call it just "renew_token"

**Pull request merged**